docs/proposals/deployment.md

```diff
@@ -0,0 +1,232 @@
+replacement depends on the DeploymentStrategy chosen by the user.
+DeploymentStrategies are explained in detail in a later section.
+
+## Implementation
+
+### API Object
+
+The `Deployment` API object will have the following structure:
+
+```go
+type Deployment struct {
+  TypeMeta
+  ObjectMeta
+
+  // Specification of the desired behavior of the Deployment.
+  Spec DeploymentSpec
+
+  // Most recently observed status of the Deployment.
+  Status DeploymentStatus
+}
+
+type DeploymentSpec struct {
+  // Number of desired pods. This is a pointer to distinguish between explicit
+  // zero and not specified. Defaults to 1.
+  Replicas *int
+
+  // Label selector for pods. Existing ReplicationControllers which have all these
+  // label selectors will be scaled down.
+  Selector map[string]string
+
+  // Describes the pods that will be created.
+  Template *PodTemplateSpec
+
+  // The deployment strategy to use to replace existing pods with new ones.
+  Strategy DeploymentStrategy
+}
+
+type DeploymentStrategy struct {
+  // Type of deployment. Can be "Recreate" or "RollingUpdate".
+  Type DeploymentType
+
+  // Rolling update config params. Present only if DeploymentType =
+  // RollingUpdate.
+  RollingUpdate *RollingUpdateDeploymentSpec
+}
+
+type DeploymentType string
+
+const (
+  // Kill all existing pods before creating new ones.
+  DeploymentRecreate DeploymentType = "Recreate"
+
+  // Replace the old RCs by new one using rolling update i.e gradually scale down the old RCs and scale up the new one.
+  DeploymentRollingUpdate DeploymentType = "RollingUpdate"
+)
+
+// Spec to control the desired behavior of rolling update.
+type RollingUpdateDeploymentSpec struct {
+  // The maximum number of pods that can be unavailable during the update.
+  // Value can be an absolute number (ex: 5) or a percentage of total pods at the start of update (ex: 10%)
+  // By default, a fixed value of 1 is used.
+  maxUnavailable IntOrString
+
+  // The maximum number of pods that be scheduled over the original number of
+  // pods.
+  // Value can be an absolute number (ex: 5) or a percentage of total pods at
+  // the start of the update (ex: 10%)
+  // By default, a value of 1 is used.
+  maxSurge IntOrString
+
+  // Key of the label that is added by the system to new pods to distinguish them from old ones.
+  // Users can set this to an empty string to indicate that the system should
+  // not add any label. If unspecified, system uses
+  // "deployment.kubernetes.io/rcHash". Value of this key is hash of the new RC.
+  uniqueLabelKey *string
+}
+
+type DeploymentStatus struct {
+  // The total number of pods created with the desired template spec.
+  Replicas int
+}
+
+```
+
+### Controller
+
+#### Deployment Controller
+
+The DeploymentController will make Deployments happen.
+It will watch Deployment objects in etcd.
+For each pending deployment, it will:
+
+1. Find all RCs whose label selector is a superset of DeploymentSpec.Selector.
```

Comment:
See my earlier comment -- you mean all RCs whose labels match DeploymentSpec.Selector, right?

If you actually mean all RCs whose Selector matches DeploymentSpec.Selector, then I'm confused. I can see that would have the effect of having the Deployment find all RCs that manage Pods that match the DeploymentSpec.Selector, but it wasn't clear to me why you want to do that instead of matching RCs. Either way, I strongly suggest explaining the motivation, since I can't be the only person who would be confused.